core/src/main/java/org/elasticsearch/cluster/routing/allocation/decider/AllocationDecider.java

```diff
@@ -98,4 +98,21 @@ public Decision canAllocate(RoutingNode node, RoutingAllocation allocation) {
     public Decision canRebalance(RoutingAllocation allocation) {
         return Decision.ALWAYS;
     }
+
+    /**
+     * Returns a {@link Decision} whether the given primary shard can be
+     * forcibly allocated on the given node. This method should only be called
+     * on nodes for which previous allocations exist for the primary shard.
```

Comment:
"This method should only be called on nodes for which previous allocations exist for the primary shard." should be more like "This method should only be called for unassigned primary shards where the node has a shard copy on disk."
Can you also assert `shardRouting.unassigned()` ?